File: packages/kits/AriaMcpKit/Tests/AriaMCPTests/AdornmentRenderTests.swift
import Testing
import Foundation
import AdornmentLib
import GeniusLocusKit
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
@testable import AriaMCP

/// Adornment render-surface tests (SPEC_ADORNMENT §5 + GENIUSLOCUSKIT_SPEC §16.2):
/// adornment text appearing in moot_memory_search payloads via the normalized
/// adornment store (ADORN-STORE-02 Part C). Tests provision minters and adornments
/// through the current store API; the retired MOOT_SUPPRESS_ADORNMENT env seam is
/// replaced by the zero-active-minters arm.
///
/// Cross-port conformance: both Swift and Rust ports read the same physical
/// fixture at Tests/Conformance/adornment_render_fixture.json for the adornment
/// text. Neither port hardcodes the string.
@Suite("Adornment render surface")
struct AdornmentRenderTests {

    // MARK: - Fixture

    /// Reads `adornment_text` from the shared cross-port fixture.
    ///
    /// Both Swift and Rust ports read the SAME physical file located at
    /// `Tests/Conformance/adornment_render_fixture.json` relative to this
    /// test file. Using the fixture rather than a hardcoded literal means a
    /// change to the fixture simultaneously breaks both ports' tests — the
    /// correct cross-port conformance signal.
    private func fixtureAdornmentText() throws -> String {
        let fixtureURL = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()  // …/AriaMCPTests
            .deletingLastPathComponent()  // …/Tests
            .appendingPathComponent("Conformance/adornment_render_fixture.json")
        let data = try Data(contentsOf: fixtureURL)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let text = json?["adornment_text"] as? String else {
            Issue.record("adornment_render_fixture.json must have an 'adornment_text' string field")
            return ""
        }
        return text
    }

    // MARK: - Harness

    /// Open an in-memory estate and return both the GeniusLocusKit handle and
    /// the underlying InMemoryStorage. The storage reference is needed by
    /// `writeAdornmentOnDrawer` to open a peer LocusKit.Estate for direct
    /// adornment writes (GLK does not expose putAdornment; the peer-estate
    /// approach is the standard test pattern for direct store writes).
    private func openEstate(
        in kit: GeniusLocusKit,
        owner: OwnerCredentials
    ) async throws -> (EstateHandle, InMemoryStorage) {
        let storage = InMemoryStorage(configuration: EstateConfiguration(
            estateID: UUID(), backend: .inMemory
        ))
        _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
        let handle = try await kit.open(
            storage: storage,
            owner: owner,
            identityKeyStore: InMemoryEstateIdentityKeyStore()
        )
        return (handle, storage)
    }

    /// Register an adornment minter and write one stored adornment for a drawer.
    ///
    /// Registration uses the GLK surface (kit.registerAdornmentMinter); the
    /// `StoredAdornment` row is written via a peer LocusKit.Estate opened on the
    /// same InMemoryStorage — the GLK actor boundary does not expose putAdornment
    /// directly. The peer-estate pattern is consistent with SearchRedactionTests
    /// and other write-side test helpers in this suite.
    ///
    /// - Parameters:
    ///   - kit: the GeniusLocusKit actor managing this estate.
    ///   - handle: the open estate handle (used for minter registration).
    ///   - minter: the descriptor to register (active flag must be true for
    ///     the adornment to appear in call-scoped active reads).
    ///   - drawerID: the drawer to adorn.
    ///   - text: the adornment text to store.
    ///   - storage: the shared InMemoryStorage (for the peer estate open).
    ///   - owner: estate owner credentials.
    /// - Returns: 1 when the adornment row was written, per `putAdornment` contract.
    @discardableResult
    private func registerMinterAndWriteAdornment(
        kit: GeniusLocusKit,
        handle: EstateHandle,
        minter: AdornmentMinterDescriptor,
        drawerID: String,
        text: String,
        storage: InMemoryStorage,
        owner: OwnerCredentials
    ) async throws -> Int {
        // Register the minter via GLK so the estate's adornment_minters table
        // carries the row. GLK delegates to Estate.registerAdornmentMinter(_:).
        try await kit.registerAdornmentMinter(in: handle, minter: minter)

        // Write the adornment row via a peer LocusKit estate on the same
        // InMemoryStorage. InMemoryStorage is a reference type — writes here
        // are immediately visible to the GLK search that follows.
        let locusEstate = try await LocusKit.Estate.open(
            storage: storage,
            owner: owner
        )
        let adornment = StoredAdornment(
            drawerID: drawerID,
            minterID: minter.id,
            text: text
        )
        return try await locusEstate.putAdornment(adornment)
    }

    /// Extract the text body from a JSONValue MCP tool result.
    private func text(of result: JSONValue) -> String {
        guard case let .object(obj) = result,
              case let .array(content)? = obj["content"],
              case let .object(first)? = content.first,
              case let .string(s)? = first["text"]
        else { return "" }
        return s
    }

    // MARK: - Tests

    /// Adornment text appears in the moot_memory_search payload for an adorned drawer.
    ///
    /// Provision path: register an active minter → write a StoredAdornment row →
    /// search returns "adornment: <text>" via the call-scoped activeAdornments read
    /// (GENIUSLOCUSKIT_SPEC §16.2).
    ///
    /// Failure mode: output is byte-identical to an un-adorned search result —
    /// no "adornment: …" line — caused by the activeAdornments batch read being
    /// absent from the candidate-list loop or the minter not being active.
    @Test("Adorned drawer with active minter surfaces 'adornment: <text>' in moot_memory_search")
    func adornmentAppearsInSearchPayload() async throws {
        let adornmentText = try fixtureAdornmentText()
        // Unique fixture query string that will match the search query.
        let fixtureQuery = "adornment-render-fixture-550e8400"

        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "adornment-render-tests-present")
        let (handle, storage) = try await openEstate(in: kit, owner: owner)
        defer { Task { try? await kit.close(handle) } }

        // Capture a drawer. Subject is set so the dense row carries the query
        // text that the search will surface.
        let frame = CaptureFrame(
            content: fixtureQuery,
            channel: .typed,
            room: "adornment-render-tests",
            latticeAnchor: .udc("004"),
            addedBy: "aria-mcp-tests",
            embeddingModelID: "test-model-v1",
            subject: fixtureQuery
        )
        let drawer = try await kit.capture(handle, frame)

        // Register an active minter and write a StoredAdornment row.
        // `isActive: true` means activeAdornments will return this row.
        let minter = AdornmentMinterDescriptor(
            id: "test-minter-apple-gen1",
            name: "Test Apple Gen-1",
            family: "apple",
            modelID: "test-model-v1",
            modelVersion: "2026-08",
            promptDigest: "sha256-test-digest-01",
            parameters: [:],
            isActive: true
        )
        let writtenCount = try await registerMinterAndWriteAdornment(
            kit: kit,
            handle: handle,
            minter: minter,
            drawerID: drawer.id,
            text: adornmentText,
            storage: storage,
            owner: owner
        )
        // Gate: putAdornment must report 1 row written.
        #expect(writtenCount == 1, "putAdornment must write exactly 1 row; got: \(writtenCount)")

        let dispatcher = ToolDispatcher(kit: kit, handle: handle)
        let result = try await dispatcher.runMemorySearch(
            ["query": .string(fixtureQuery)]
        )
        let body = text(of: result)

        // S2 row format: id · subject · firstSentence · SSC · <adornmentText> · eventTime
        // The adornment text is column 5 (no label prefix) per ARIA_MCP_INTERFACE.md §11.5.
        #expect(
            body.contains(" · \(adornmentText) · "),
            "adorned drawer must surface adornment text as column 5 of the S2 row; got: \(body)"
        )
    }

    /// Zero active minters → no adornment line in moot_memory_search.
    ///
    /// The zero-active-minters arm replaces the retired MOOT_SUPPRESS_ADORNMENT
    /// env seam (ADORN-STORE-02 Part C). When no minters are active, the
    /// call-scoped activeAdornments read returns an empty map and no
    /// "adornment: …" line is appended to the payload.
    ///
    /// Failure mode: "adornment: …" still appears in output — caused by the
    /// activeAdornments map not being consulted (falling back to the retired
    /// drawer.adornment field read).
    @Test("Zero active minters → no adornment line in moot_memory_search")
    func zeroActiveMintersSuppressesAdornmentLine() async throws {
        let fixtureQuery = "adornment-suppress-fixture-550e8401"

        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "adornment-render-tests-suppress")
        let (handle, _) = try await openEstate(in: kit, owner: owner)
        defer { Task { try? await kit.close(handle) } }

        // Capture a drawer with a unique query string.
        let frame = CaptureFrame(
            content: fixtureQuery,
            channel: .typed,
            room: "adornment-suppress-tests",
            latticeAnchor: .udc("004"),
            addedBy: "aria-mcp-tests",
            embeddingModelID: "test-model-v1",
            subject: fixtureQuery
        )
        _ = try await kit.capture(handle, frame)

        // No minters registered → activeAdornments returns empty map
        // → no "adornment:" line in the search payload.
        let dispatcher = ToolDispatcher(kit: kit, handle: handle)
        let result = try await dispatcher.runMemorySearch(
            ["query": .string(fixtureQuery)]
        )
        let body = text(of: result)

        #expect(
            !body.contains("adornment:"),
            "zero active minters must produce no adornment line; got: \(body)"
        )
    }

    /// Tool schema strings must not mention MOOT_SUPPRESS_ADORNMENT.
    ///
    /// This is a regression guard: the suppression seam was retired in
    /// ADORN-STORE-02 Part C and must not re-appear in AI-visible tool
    /// descriptions or inputSchema JSON. If it leaks, any LLM client that
    /// reads tool descriptions learns about the seam and could attempt to
    /// control benchmark conditions.
    @Test("Tool schemas do not mention MOOT_SUPPRESS_ADORNMENT")
    func toolSchemasDoNotMentionSuppressVar() throws {
        // Use empty environment to bypass vault/memory gating and get all tools.
        // The point is to check every schema string, not to test gate logic.
        let tools = ToolProjection.tools(environment: [:])
        for tool in tools {
            #expect(
                !tool.description.contains("MOOT_SUPPRESS_ADORNMENT"),
                "tool \(tool.name) description must not mention MOOT_SUPPRESS_ADORNMENT"
            )
            // Encode the inputSchema to JSON text for substring search.
            let schemaText: String
            if let data = try? tool.inputSchema.encoded(),
               let s = String(data: data, encoding: .utf8) {
                schemaText = s
            } else {
                schemaText = ""
            }
            #expect(
                !schemaText.contains("MOOT_SUPPRESS_ADORNMENT"),
                "tool \(tool.name) inputSchema must not mention MOOT_SUPPRESS_ADORNMENT"
            )
        }
    }
}

File: packages/kits/AriaMcpKit/rust/tests/adornment_render_tests.rs
//! Adornment render-surface tests (ADORN-STORE-02 Part C):
//! adornment text appearing in moot_memory_search payloads when provisioned
//! via the normalized adornments store, and zero-active-minters as the
//! suppression arm.
//!
//! Cross-port conformance: reads the SAME physical fixture as the Swift tests
//! at `Tests/Conformance/adornment_render_fixture.json`. Neither port hardcodes
//! the adornment text — changing the fixture simultaneously breaks both ports.
//!
//! Provisioning path (ADORN-STORE-02): `register_adornment_minter` +
//! `put_adornment` via `DrawerStore` trait on the registry's store. The retired
//! `set_adornment` + `MOOT_SUPPRESS_ADORNMENT` seam is gone; zero-active-minters
//! is now the suppression arm (no active minters → no adornment line in the
//! search payload).

use std::collections::BTreeMap;
use std::path::Path;

use adornment_lib::{AdornmentMinterDescriptor, StoredAdornment};
use aria_mcp::{
    dispatch::dispatch_tool,
    estate_registry::EstateRegistry,
    jsonrpc::JsonValue,
    surfaced_recall_ledger::SurfacedRecallLedger,
    tool_list::build_tool_list_with_flags,
};
use locus_kit::drawer_store::DrawerStore;

// ---------------------------------------------------------------------------
// Fixture + test helpers
// ---------------------------------------------------------------------------

/// Read the shared `adornment_render_fixture.json` and return `adornment_text`.
///
/// CARGO_MANIFEST_DIR resolves to `packages/kits/AriaMcpKit/rust/`.
/// The fixture lives at `packages/kits/AriaMcpKit/Tests/Conformance/`, i.e.,
/// one level up from `rust/` then into `Tests/Conformance/`.
fn fixture_adornment_text() -> String {
    let manifest_dir = std::env::var("CARGO_MANIFEST_DIR")
        .expect("CARGO_MANIFEST_DIR must be set during cargo test");
    let fixture_path = Path::new(&manifest_dir)
        .parent()   // AriaMcpKit/
        .expect("parent of rust/ must exist")
        .join("Tests/Conformance/adornment_render_fixture.json");

    let data = std::fs::read_to_string(&fixture_path)
        .unwrap_or_else(|e| {
            panic!("Failed to read fixture at {}: {}", fixture_path.display(), e)
        });

    let parsed: serde_json::Value = serde_json::from_str(&data)
        .expect("adornment_render_fixture.json must be valid JSON");

    parsed
        .get("adornment_text")
        .and_then(|v| v.as_str())
        .expect("adornment_render_fixture.json must have an 'adornment_text' string field")
        .to_owned()
}

macro_rules! args {
    () => { BTreeMap::new() };
    ( $( $k:expr => $v:expr ),+ $(,)? ) => {{
        let mut m = BTreeMap::new();
        $( m.insert($k.to_string(), JsonValue::from(serde_json::json!($v))); )+
        m
    }};
}

fn content_text(result: &serde_json::Value) -> &str {
    result["content"][0]["text"].as_str().unwrap_or("")
}

fn is_success(result: &serde_json::Value) -> bool {
    result["isError"] == serde_json::json!(false)
}

/// File a drawer with content + subject and return its ID.
fn file_memory(registry: &EstateRegistry, content: &str, location: &str) -> String {
    let subject: String = content.chars().take(120).collect();
    let a = args![
        "content" => content,
        "location" => location,
        "subject" => subject.as_str(),
        "impatient" => true
    ];
    let result = dispatch_tool(
        "moot_file_memory",
        &a,
        registry,
        &SurfacedRecallLedger::new(),
    )
    .expect("moot_file_memory must succeed");
    assert!(is_success(&result), "file_memory should succeed; got: {result:?}");
    let text = content_text(&result);
    // Result line: "filed memory <id>\nroom: ..."
    text.lines()
        .next()
        .and_then(|l| l.strip_prefix("filed memory "))
        .unwrap_or("")
        .to_owned()
}

/// Register a test minter and write one adornment row for `drawer_id` via the
/// normalized adornments store (ADORN-STORE-02). Returns the minter ID so the
/// caller can deactivate it to test the zero-active-minters suppression arm.
fn register_minter_and_write_adornment(
    registry: &EstateRegistry,
    minter_id: &str,
    drawer_id: &str,
    text: &str,
) {
    // is_active = true: active_adornments() returns this row in the search path.
    let minter = AdornmentMinterDescriptor::new(
        minter_id,
        "Test Minter",
        "test",
        "test-model-v1",
        "2026",
        "test-prompt-digest",
        BTreeMap::new(),
        true,
    );
    registry
        .default
        .store
        .register_adornment_minter(&minter)
        .unwrap_or_else(|e| panic!("register_adornment_minter must succeed: {e}"));

    let adornment = StoredAdornment::new(drawer_id, minter_id, text);
    registry
        .default
        .store
        .put_adornment(&adornment)
        .unwrap_or_else(|e| panic!("put_adornment must succeed: {e}"));
}

// ---------------------------------------------------------------------------
// Test 1: adornment appears in search payload when active minter is present
// ---------------------------------------------------------------------------

/// Active minter + adornment row → the adornment text surfaces as column 5 of
/// the S1 row in the search payload. Provisioned via the normalized adornments
/// store (ADORN-STORE-02).
///
/// Failure mode: the adornment text absent from the payload — `active_adornments`
/// batch read missing from the search render loop, or minter not active.
#[test]
fn adornment_appears_in_search_payload() {
    let adornment_text = fixture_adornment_text();
    let fixture_query = "adornment-render-fixture-store-rust";

    let registry = EstateRegistry::new_inmemory_bare();
    let drawer_id =
        file_memory(&registry, fixture_query, "adornment-tests");
    register_minter_and_write_adornment(
        &registry,
        "test-minter-active-render",
        &drawer_id,
        &adornment_text,
    );

    let result = dispatch_tool(
        "moot_memory_search",
        &args!["query" => fixture_query],
        &registry,
        &SurfacedRecallLedger::new(),
    )
    .expect("moot_memory_search must not throw");

    assert!(is_success(&result), "search must succeed; got: {result:?}");
    let body = content_text(&result);
    // S1 row format: id · subject · firstSentence · SSC · <adornmentText> · eventTime · score
    // The adornment text is column 5 (no label prefix) per ARIA_MCP_INTERFACE.md §11.2,
    // the same assertion AdornmentRenderTests.swift makes.
    let expected_column = format!(" · {} · ", adornment_text);
    assert!(
        body.contains(&expected_column),
        "adorned drawer must surface adornment text as column 5 of the S1 row;\n\
         expected column: {expected_column}\n\
         got: {body}"
    );
}

// ---------------------------------------------------------------------------
// Test 2: zero active minters → adornment line absent (suppression arm)
// ---------------------------------------------------------------------------

/// No active minters → adornment line absent from the search payload.
/// Zero-active-minters IS the suppression arm (replaces retired MOOT_SUPPRESS_ADORNMENT).
///
/// Failure mode: the adornment text still appears — active_adornments batch read
/// not filtering by active minter, or old d.adornment path still live.
#[test]
fn zero_active_minters_suppresses_adornment_line() {
    let adornment_text = fixture_adornment_text();
    let fixture_query = "adornment-render-zero-minters-rust";

    let registry = EstateRegistry::new_inmemory_bare();
    // File the memory but do NOT register any active minters or write any adornment rows.
    // The search result must have no "adornment:" line.
    let _drawer_id = file_memory(&registry, fixture_query, "adornment-tests");

    let result = dispatch_tool(
        "moot_memory_search",
        &args!["query" => fixture_query],
        &registry,
        &SurfacedRecallLedger::new(),
    )
    .expect("moot_memory_search must not throw");

    assert!(is_success(&result), "search must succeed; got: {result:?}");
    let body = content_text(&result);
    // No adornment rows in the store → column 5 renders '-' and the text is absent.
    assert!(
        !body.contains(&adornment_text),
        "no active minters → adornment text must be absent from payload;\n\
         got: {body}\n\
         (fixture adornment_text was: {adornment_text})"
    );
}

// ---------------------------------------------------------------------------
// Test 3: tool schema JSON does not mention MOOT_SUPPRESS_ADORNMENT
// ---------------------------------------------------------------------------

/// The retired MOOT_SUPPRESS_ADORNMENT seam must not appear in any tool
/// description or inputSchema JSON — it is a benchmark-internal identifier
/// and must never be exposed to AI clients via the tool catalog.
///
/// Failure mode: "MOOT_SUPPRESS_ADORNMENT" appears in a tool description or
/// inputSchema — the old env-var name leaked into public-facing schema text.
#[test]
fn tool_schemas_do_not_mention_suppress_var() {
    // vault_on=true, memory_on=false: full tool surface without the opt-in
    // memory-adapter tier; gives the widest tool list to check.
    let tools_json = build_tool_list_with_flags(true, false);
    let tools_str = tools_json.to_string();

    assert!(
        !tools_str.contains("MOOT_SUPPRESS_ADORNMENT"),
        "tool list JSON must not mention MOOT_SUPPRESS_ADORNMENT — \
         seam is retired (ADORN-STORE-02); found in tool list: {}",
        if tools_str.contains("MOOT_SUPPRESS_ADORNMENT") { "yes" } else { "no" }
    );
}
